src/webutil/tableState.js

```diff
@@ -0,0 +1,236 @@
+    initialRowsPerPage?: number,
+    // Defaults to unsorted (maintains order of the original data).
+    initialSort?: {
+      // The starting order of the initial sort, represented by "asc" or "desc".
+      sortOrder: SortOrderTypes,
+      // A key representing the sort field. It is is used to toggle sort order and increase efficiency.
+      sortName: string,
+      // A function that returns the property of the element that should be used in the sort comparator.
+      sortFn: PluckFn<T>,
+    },
+  }
+): TableState<T> {
+  const optionDefaults = {
+    initialRowsPerPage: data.length,
+    initialSort: {},
+    ...initialOptions,
+  };
+  const sortDefaults = {
+    sortOrder: SortOrders.ASC,
+    sortName: "",
+    sortFn: null,
+    ...optionDefaults.initialSort,
+  };
+  const [sortName, setSortName] = useState(sortDefaults.sortName);
+  const [sortOrder, setSortOrder] = useState<SortOrderTypes>(
+    sortDefaults.sortOrder
+  );
+  const [sortFn, setSortFn] = useState(() => sortDefaults.sortFn);
+  const [pageIndex, setPageIndex] = useState(0);
+  const [rowsPerPage, setRowsPerPageInternal] = useState(
+    optionDefaults.initialRowsPerPage
+  );
+  type FilterFns = {[string]: null | ((T) => boolean)};
+  const [filterFns, setFilterFns] = useState<FilterFns>({});
+
+  const sortedArray = useMemo(() => {
+    const array = data.slice();
+    if (sortFn) {
+      sortBy(array, sortFn);
+      if (sortOrder === SortOrders.DESC) array.reverse();
+    }
+    return array;
+  }, [data, sortFn, sortOrder]);
+
+  const filteredArray = useMemo(() => {
+    const reducer = (array, key) => {
+      const fn = filterFns[key];
+      if (fn) return array.filter(fn);
+      return array;
+    };
+    return Object.keys(filterFns).reduce(reducer, sortedArray);
+  }, [sortedArray, filterFns]);
+
+  const setRowsPerPage = useCallback((newRowsPerPage: number) => {
+    // Keep the user from losing their place when changing # of results per page
+    const intRowsPerPage = parseInt(newRowsPerPage, 10);
+    const currentFirstVisibleRow = pageIndex * rowsPerPage;
+    const newPage = Math.floor(currentFirstVisibleRow / intRowsPerPage);
+
+    setRowsPerPageInternal(intRowsPerPage);
+    setPageIndex(newPage);
+  });
+
+  const setSortFnCallback = useCallback(
+    (newSortName: string, newSortFn: PluckFn<T>) => {
+      if (newSortName === sortName) {
+        setSortOrder(
+          sortOrder === SortOrders.DESC ? SortOrders.ASC : SortOrders.DESC
+        );
+      } else {
+        setSortOrder(DEFAULT_SORT);
+        setSortFn(() => newSortFn);
+      }
+      setSortName(newSortName);
+    }
+  );
+
+  const createOrUpdateFilterFn = useCallback(
+    (key: string, newFilterFn: null | ((T) => boolean)) => {
+      const setterFn: (FilterFns) => FilterFns = (fns) => {
+        return {
+          ...fns,
+          [key]: newFilterFn,
+        };
+      };
+      setPageIndex(0);
+      setFilterFns(setterFn);
+    }
+  );
```

Comment:
Although this one doesn't have any dependencies on values outside the callback, you still need to pass an empty array as the second param to useCallback in order to indicate that

Reply:
sure